# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

## Задание 1
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [2]:
import pandas as pd

In [66]:
hd = pd.read_csv('horse_data.csv', 
                 usecols = [0, 1, 2, 3, 4, 6, 7, 8], 
                 names = ['Surgery', 'Age', 'Hospital ID', 'Rectal temp', 'Pulse', 'Temp of extremities', 'Peripheral pulse', 'Mucous membranes'],
                 na_values = "?")
hd.head()

Surgery  Age  Hospital ID  Rectal temp  Pulse  Temp of extremities  \
0      2.0    1       530101         38.5   66.0                  3.0   
1      1.0    1       534817         39.2   88.0                  NaN   
2      2.0    1       530334         38.3   40.0                  1.0   
3      1.0    9      5290409         39.1  164.0                  4.0   
4      2.0    1       530255         37.3  104.0                  NaN   

   Peripheral pulse  Mucous membranes  
0               3.0               NaN  
1               NaN               4.0  
2               1.0               3.0  
3               1.0               6.0  
4               NaN               6.0

In [78]:
# чтобы можно было работать с NaN-значениями в целочисленных столбцах
hd[['Surgery', 'Temp of extremities', 'Peripheral pulse', 'Hospital ID', 'Mucous membranes']] \
= hd[['Surgery', 'Temp of extremities', 'Peripheral pulse', 'Hospital ID', 'Mucous membranes']].astype('Int64')
hd.head()

Surgery  Age  Hospital ID  Rectal temp  Pulse  Temp of extremities  \
0        2    1       530101         38.5   66.0                    3   
1        1    1       534817         39.2   88.0                 <NA>   
2        2    1       530334         38.3   40.0                    1   
3        1    9      5290409         39.1  164.0                    4   
4        2    1       530255         37.3  104.0                 <NA>   

   Peripheral pulse  Mucous membranes  
0                 3              <NA>  
1              <NA>                 4  
2                 1                 3  
3                 1                 6  
4              <NA>                 6

In [79]:
# в сопроводительном документе указан диапазон значений [1, 2] для столбца 'Age'.
# в csv-файле вместо 2 значение 9. Заменим в датафрейме 9 на 2 для удобства восприятия
hd['Age'] = hd['Age'].apply(lambda x: 2 if x == 9 else x)
hd.head()

Surgery  Age  Hospital ID  Rectal temp  Pulse  Temp of extremities  \
0        2    1       530101         38.5   66.0                    3   
1        1    1       534817         39.2   88.0                 <NA>   
2        2    1       530334         38.3   40.0                    1   
3        1    2      5290409         39.1  164.0                    4   
4        2    1       530255         37.3  104.0                 <NA>   

   Peripheral pulse  Mucous membranes  
0                 3              <NA>  
1              <NA>                 4  
2                 1                 3  
3                 1                 6  
4              <NA>                 6

In [26]:
hd.describe()

Surgery         Age   Hospital ID  Rectal temp       Pulse  \
count  299.000000  300.000000  3.000000e+02   240.000000  276.000000   
mean     1.397993    1.080000  1.085889e+06    38.167917   71.913043   
std      0.490305    0.271746  1.529801e+06     0.732289   28.630557   
min      1.000000    1.000000  5.184760e+05    35.400000   30.000000   
25%      1.000000    1.000000  5.289040e+05    37.800000   48.000000   
50%      1.000000    1.000000  5.303055e+05    38.200000   64.000000   
75%      2.000000    1.000000  5.347275e+05    38.500000   88.000000   
max      2.000000    2.000000  5.305629e+06    40.800000  184.000000   

       Temp of extremities  Peripheral pulse  Mucous membranes  
count           244.000000        231.000000        253.000000  
mean              2.348361          2.017316          2.853755  
std               1.045054          1.042428          1.620294  
min               1.000000          1.000000          1.000000  
25%               1.000000          1.000000          1.000000  
50%               3.000000          2.000000          3.000000  
75%               3.000000          3.000000          4.000000  
max               4.000000          4.000000          6.000000

Первичные выводы:
- В то время как данные по хирургическому вмешательству, возрасту и больничному номеру есть для всех пациентов (только для одной лошади отсутствуют данные по факту проведения операции), замеры показателей жизненно важных функций проводились не поголовно;
- Из столбца Surgery по среднему значению видим, что операция проводилась около 40% пациентов;
- Взрослых лошадей в базе намного больше, чем жеребят (по среднему значению получаем - 92%);
- Ректальная температура в среднем выше нормы (38.2 при норме 37.8). СКО составляет 0,73 градуса Цельсия;
- Cредний пульс намного выше нормы: 72 уд/мин при норме 30-40. Менее 25% лошадей имеют нормальный или близкий к нормальному пульс;
- У около 50% пациентов охлаждённые конечности, что может свидетельствовать о состоянии шока и сниженном кровоснабжении. Указано, что повышенная температура конечностей может свидетельствовать о повышенной ректальной температуре. Проверим это:

In [33]:
hd.loc[hd['Temp of extremities'] == 2]['Rectal temp'].mean()

38.20384615384615

Действительно, у лошадей с горячими конечностями наблюдается повышенная ректальная температура.

## Задание 2
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [52]:
# начнём проводить расчёт выбросов для каждого числового столбца на основе межквартильного размаха
# ректальная температура
rt_q1 = hd['Rectal temp'].quantile(0.25)
rt_q3 = hd['Rectal temp'].quantile(0.75)
rt_iqr = rt_q3 - rt_q1
rt_lb = rt_q1 - (1.5 * rt_iqr)
rt_ub = rt_q3 + (1.5 * rt_iqr)
hd['Rectal temp'][(~hd['Rectal temp'].between(rt_lb, rt_ub, inclusive = True))].dropna()

20     39.9
44     35.4
54     40.3
75     39.7
80     36.4
91     40.3
99     39.6
118    36.5
141    36.0
238    36.1
251    36.6
259    40.8
281    40.0
298    36.5
Name: Rectal temp, dtype: float64

Найденные по формуле значения сложно отнести к выбросам. Скорее всего, это просто индикаторы особо болезненных состояний пациентов. Эти значения следует оставить в выборке.

In [57]:
# пульс
p_q1 = hd['Pulse'].quantile(0.25)
p_q3 = hd['Pulse'].quantile(0.75)
p_iqr = p_q3 - p_q1
p_lb = p_q1 - (1.5 * p_iqr)
p_ub = p_q3 + (1.5 * p_iqr)
hd['Pulse'][(~hd['Pulse'].between(p_lb, p_ub, inclusive = True))].dropna()

3      164.0
41     150.0
55     160.0
255    184.0
275    150.0
Name: Pulse, dtype: float64

То же самое, что и в предыдущем случае: полученные по формуле значения не выбиваются из общего ряда наблюдений и могут быть экстремальными состояниями пациентов. Эти значения тоже следует оставить.

## Задание 3
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [67]:
hd.isna().sum()

Surgery                 1
Age                     0
Hospital ID             0
Rectal temp            60
Pulse                  24
Temp of extremities    56
Peripheral pulse       69
Mucous membranes       47
dtype: int64

- Для столбцов 1-3 пропусков нет, отсутствует лишь одна запись о хирургическом вмешательстве для одной лошади. Поскольку заполнение её любым значением не повлияет на общую картину, можно проигнорировать эту запись или заполнить любым значением на выбор;
- В столбце Rectal temp 60 пропусков. Поскольку данные отсутствуют в случайном порядке и процент пропусков невелик, можно исключить строки с пропусками или заполнить их средним. Это не повлияет существенным образом на характеристики, лишь немного изменится СКО и медиана:


In [70]:
hd['Rectal temp'].dropna().describe()

count    240.000000
mean      38.167917
std        0.732289
min       35.400000
25%       37.800000
50%       38.200000
75%       38.500000
max       40.800000
Name: Rectal temp, dtype: float64

In [71]:
hd['Rectal temp'].fillna(hd['Rectal temp'].mean()).describe()

count    300.000000
mean      38.167917
std        0.654705
min       35.400000
25%       37.900000
50%       38.167917
75%       38.500000
max       40.800000
Name: Rectal temp, dtype: float64

- В столбце Pulse 24 пропуска. Поступаем так же, как и с предыдущим столбцом:

In [72]:
hd['Pulse'].dropna().describe()

count    276.000000
mean      71.913043
std       28.630557
min       30.000000
25%       48.000000
50%       64.000000
75%       88.000000
max      184.000000
Name: Pulse, dtype: float64

In [73]:
hd['Pulse'].fillna(hd['Pulse'].mean()).describe()

count    300.000000
mean      71.913043
std       27.457472
min       30.000000
25%       48.000000
50%       68.000000
75%       88.000000
max      184.000000
Name: Pulse, dtype: float64

- Столбцы Temp of extremities, Peripheral pulse и Mucous membranes содержат категориальные значения. В данном случае заполнять пропуски средним не имеет смысла. Примем, что отсутствие данных о признаке болезни означает его отсутствие. 

In [89]:
hd['Surgery'] = hd['Surgery'].fillna(2)
hd['Rectal temp'] = hd['Rectal temp'].fillna(hd['Rectal temp'].mean())
hd['Pulse'] = hd['Pulse'].fillna(hd['Pulse'].mean())
hd['Temp of extremities'] = hd['Temp of extremities'].fillna(1)
hd['Peripheral pulse'] = hd['Peripheral pulse'].fillna(1)
hd['Mucous membranes'] = hd['Mucous membranes'].fillna(1)
hd

Surgery  Age  Hospital ID  Rectal temp  Pulse  Temp of extremities  \
0          2    1       530101    38.500000   66.0                    3   
1          1    1       534817    39.200000   88.0                    1   
2          2    1       530334    38.300000   40.0                    1   
3          1    2      5290409    39.100000  164.0                    4   
4          2    1       530255    37.300000  104.0                    1   
..       ...  ...          ...          ...    ...                  ...   
295        1    1       533886    38.167917  120.0                    4   
296        2    1       527702    37.200000   72.0                    3   
297        1    1       529386    37.500000   72.0                    4   
298        1    1       530612    36.500000  100.0                    3   
299        1    1       534618    37.200000   40.0                    1   

     Peripheral pulse  Mucous membranes  
0                   3                 1  
1                   1                 4  
2                   1                 3  
3                   1                 6  
4                   1                 6  
..                ...               ...  
295                 1                 4  
296                 2                 4  
297                 3                 4  
298                 3                 3  
299                 1                 1  

[300 rows x 8 columns]

In [90]:
hd.isna().sum()

Surgery                0
Age                    0
Hospital ID            0
Rectal temp            0
Pulse                  0
Temp of extremities    0
Peripheral pulse       0
Mucous membranes       0
dtype: int64

## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

#### ПРИМЕЧАНИЕ
Домашнее задание сдается ссылкой на репозиторий [GitHub](https://github.com/).
Не сможем проверить или помочь, если вы пришлете:
- файлы;
- архивы;
- скриншоты кода.

Все обсуждения и консультации по выполнению домашнего задания ведутся только на соответствующем канале в slack.

##### Как правильно задавать вопросы аспирантам, преподавателям и коллегам
Прежде чем задать вопрос, попробуйте найти ответ в интернете. Навык самостоятельного поиска информации — один из важнейших. Каждый практикующий специалист любого уровня делает это ежедневно.

Сформулируйте вопрос по алгоритму:  
1) Что я делаю?  
2) Какого результата я ожидаю?  
3) Как фактический результат отличается от ожидаемого?  
4) Что я уже попробовал сделать, чтобы исправить проблему?  

По возможности прикрепите к вопросу скриншоты либо ссылки на код. Не выкладывайте все решение, оставляйте только проблемный и воспроизводимый участок кода.